android/src/main/java/com/tailscale/ipn/NetworkChangeCallback.kt

```diff
@@ -6,94 +6,159 @@ import android.net.ConnectivityManager
-                capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)) {
-              maybeUpdateDNSConfig(linkProperties, dns)
-            } else {
-              if (!capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_VPN) ||
-                  !capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)) {
-                Log.d(
-                    TAG,
-                    "Capabilities changed for network $network.toString(), but not updating DNS config because either this is a VPN network or non-internet network")
-              } else {
-                Log.d(
-                    TAG,
-                    "Capabilities changed for network $network.toString(), but not updating DNS config, because the LinkProperties hasn't been gotten yet")
-              }
+            lock.withLock {
+              activeNetworks[network]?.caps = capabilities
+              maybeUpdateDNSConfigLocked("onCapabilitiesChanged", dns)
             }
           }
 
           override fun onLinkPropertiesChanged(network: Network, linkProperties: LinkProperties) {
             super.onLinkPropertiesChanged(network, linkProperties)
-            linkPropertiesCache[network] = linkProperties
-            val capabilities = networkCapabilitiesCache[network]
-            if (capabilities != null &&
-                capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_VPN) &&
-                capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)) {
-              maybeUpdateDNSConfig(linkProperties, dns)
-            } else {
-              if (capabilities == null) {
-                Log.d(
-                    TAG,
-                    "Capabilities changed for network $network.toString(), but not updating DNS config because capabilities haven't been gotten for this network yet")
-              } else {
-                Log.d(
-                    TAG,
-                    "Capabilities changed for network $network.toString(), but not updating DNS config, because this is a VPN network or non-Internet network")
-              }
+            lock.withLock {
+              activeNetworks[network]?.linkProps = linkProperties
+              maybeUpdateDNSConfigLocked("onLinkPropertiesChanged", dns)
             }
           }
 
           override fun onLost(network: Network) {
             super.onLost(network)
-            if (dns.updateDNSFromNetwork("")) {
-              Libtailscale.onDNSConfigChanged("")
+
+            Log.d(TAG, "onLost: network ${network.toString()}")
+            lock.withLock {
+              activeNetworks.remove(network)
+              maybeUpdateDNSConfigLocked("onLost", dns)
             }
           }
         })
   }
 
-  fun maybeUpdateDNSConfig(linkProperties: LinkProperties, dns: DnsConfig) {
+  // pickNonMetered returns the first non-metered network in the list of
+  // networks, or the first network if none are non-metered.
```

Comment:
Can there be a situation where there are multiple non-metered networks, so picking the first one might result in the wrong one being selected?

Reply:
Sorry, we discussed this earlier. Resolving.

Reply:
It is possible, but should be ok to just pick one.